File: src/main/java/net/hectus/neobb/player/NeoInventory.java
package net.hectus.neobb.player;

import com.marcpg.libpg.lang.Translation;
import com.marcpg.libpg.util.ItemBuilder;
import com.marcpg.libpg.util.Randomizer;
import net.hectus.neobb.turn.Turn;
import net.hectus.neobb.turn.person_game.categorization.Category;
import org.bukkit.Material;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.PlayerInventory;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.stream.IntStream;

public class NeoInventory {
    private NeoPlayer player;
    private final ItemStack[] deck;
    private final Turn<?>[] dummyTurnDeck;
    private int coins;
    private boolean shopDone;

    public NeoInventory(@NotNull NeoPlayer player) {
        this.player = player;
        this.coins = player.game.info().coins();
        this.deck = new ItemStack[player.game.info().deckSize()];
        this.dummyTurnDeck = new Turn<?>[player.game.info().deckSize()];
        sync();
    }

    public void switchOwner(NeoPlayer player) {
        this.player = player;
    }

    public void clear() {
        for (int i = 0; i < player.game.info().deckSize(); i++) {
            this.deck[i] = null;
            this.dummyTurnDeck[i] = null;
        }
        sync();
    }

    public boolean isEmpty() {
        for (ItemStack i : deck) {
            if (i != null) return false;
        }
        return true;
    }

    public boolean isFull() {
        for (ItemStack i : deck) {
            if (i == null) return false;
        }
        return true;
    }

    public ItemStack[] deck() {
        return deck;
    }

    public Turn<?>[] dummyTurnDeck() {
        return dummyTurnDeck;
    }

    public void setDeckSlot(int slot, @Nullable ItemStack item, @Nullable Turn<?> turn) {
        this.deck[slot] = item;
        this.dummyTurnDeck[slot] = turn;
        sync();
    }

    public void addToDeck(@NotNull Turn<?> turn) {
        turn.items().forEach(item -> addToDeck(new ItemBuilder(item)
                .lore(player.shop.loreBuilder.turn(turn).buildWithTooltips(player.locale()))
                .build(), turn));
    }

    public void addToDeck(ItemStack item, Turn<?> turn) {
        for (int i = 0; i < deck.length; i++) {
            if (deck[i] == null) {
                setDeckSlot(i, item, turn);
                return;
            }
        }
        throw new ArrayIndexOutOfBoundsException("Deck is already full!");
    }

    public void addRandom() {
        if (isFull()) return;
        try {
            addToDeck(Randomizer.fromCollection(player.shop.dummyTurns));
        } catch (ArrayIndexOutOfBoundsException ignored) {} // Skill issue I guess.
    }

    public void removeRandom() {
        if (isEmpty()) return;
        setDeckSlot(Randomizer.fromCollection(IntStream.range(0, deck.length)
                .filter(i -> deck[i] != null)
                .boxed()
                .toList()
        ), null, null);
    }

    public void fillInRandomly() {
        for (ItemStack item : deck) {
            if (item == null) addRandom();
        }
    }

    public boolean allowItem(Turn<?> turn, Material material) {
        int count = 0;
        for (ItemStack item : deck) {
            if (item != null && item.getType() == material) count++;
        }
        return count < (turn instanceof Category personTurn ? personTurn.categoryMaxPerDeck() :  turn.maxAmount());
    }

    public synchronized boolean removeCoins(int coins) {
        if (this.coins < coins) return false;
        this.coins -= coins;
        player.inventory.sync();
        return true;
    }

    public synchronized void addCoins(int coins) {
        this.coins += coins;
        player.inventory.sync();
    }

    public boolean shopDone() {
        return shopDone;
    }

    public void setShopDone(boolean shopDone) {
        this.shopDone = shopDone;
    }

    public void sync() {
        PlayerInventory inv = player.player.getInventory();
        inv.clear();
        for (int i = 0; i < deck.length; i++) {
            ItemStack item = deck[i] == null ? new ItemStack(Material.BLACK_STAINED_GLASS_PANE) : deck[i].clone();
            Turn<?> dummyTurn = dummyTurnDeck[i] == null ? Turn.DUMMY : dummyTurnDeck[i];

            if (player.game.difficulty.suggestions && dummyTurn.goodChoice(player))
                item.addUnsafeEnchantment(Enchantment.MENDING, 1);

            inv.setItem(i, item);
        }
        inv.setItem(13, new ItemBuilder(coins == 0 ? Material.REDSTONE : Material.GOLD_INGOT)
                .amount(Math.clamp(1, coins, 64))
                .name(Translation.component(player.locale(), "item-lore.cost.value", coins))
                .build());
    }
}

File: src/main/kotlin/net/hectus/neobb/player/NeoInventory.kt
package net.hectus.neobb.player

import com.marcpg.libpg.item.ItemBuilder
import com.marcpg.libpg.util.asString
import com.marcpg.libpg.util.component
import net.hectus.neobb.modes.shop.util.Items
import net.hectus.neobb.modes.turn.Turn
import net.hectus.neobb.modes.turn.person_game.Category
import org.bukkit.Material
import org.bukkit.inventory.ItemStack
import java.util.stream.IntStream

class NeoInventory(private var player: NeoPlayer) {
    var coins: Int = player.game.info.coins
    var shopDone: Boolean = false

    val deck: Array<ItemStack?> = arrayOfNulls(player.game.info.deckSize)
    val turns: Array<Turn<*>?> = arrayOfNulls(player.game.info.deckSize)

    init {
        sync()
    }

    fun switchOwner(newOwner: NeoPlayer) {
        player = newOwner
    }

    fun clear() {
        for (i in deck.indices) {
            deck[i] = null
            turns[i] = null
        }
        sync()
    }

    fun isEmpty(): Boolean = deck.all { it == null }
    fun isFull(): Boolean = deck.all { it != null }
    fun slotsLeft(): Int = deck.count { it == null }

    fun setSlot(slot: Int, item: ItemStack?, turn: Turn<*>?) {
        deck[slot] = item
        turns[slot] = turn
        sync()
    }

    fun removeOne(item: ItemStack) {
        val index = deck.indexOfLast { item.type == it?.type && item.displayName().asString() == it.displayName().asString() }
        if (index == -1) return
        deck[index] = null
        turns[index] = null
    }

    fun sell(turn: Turn<*>) {
        turn.items.forEach { removeOne(it) }
        addCoins(turn.cost ?: 0)
    }

    fun clearSlot(slot: Int) {
        setSlot(slot, null, null)
    }

    fun clearFirst(predicate: (ItemStack, Turn<*>) -> Boolean) {
        for (i in deck.indices) {
            if (deck[i] != null && predicate(deck[i]!!, turns[i]!!)) {
                clearSlot(i)
                return
            }
        }
    }

    fun add(turn: Turn<*>) {
        turn.items.forEach { i -> add(player.shop.makeItem(turn, i, false), turn) }
    }

    fun add(item: ItemStack, turn: Turn<*>) {
        for (i in deck.indices) {
            if (deck[i] == null) {
                setSlot(i, item, turn)
                return
            }
        }
        throw ArrayIndexOutOfBoundsException("Deck is already full!")
    }

    fun addRandom() {
        if (isFull()) return

        runCatching {
            add(player.game.info.turns.random())
        }
    }

    fun removeRandom() {
        if (isEmpty()) return

        setSlot(IntStream.range(0, deck.size)
            .filter { i -> deck[i] != null }
            .boxed()
            .toList().random(), null, null)
    }

    fun fillInRandomly() {
        for (item in deck) {
            if (item == null) addRandom()
        }
    }

    fun allows(turn: Turn<*>): Boolean {
        var count = 0
        for (item in deck) {
            if (item != null && item.type == turn.mainItem.type)
                count++
        }
        return count < (if (turn is Category) turn.categoryMaxPerDeck else turn.maxAmount)
    }

    @Synchronized
    fun addCoins(coins: Int) {
        this.coins += coins
        sync()
    }

    @Synchronized
    fun removeCoins(coins: Int): Boolean {
        if (coins > this.coins) return false
        this.coins -= coins
        sync()
        return true
    }

    fun sync() {
        val inv = player.player.inventory
        inv.clear()

        for (i in deck.indices) {
            val item = if (deck[i] == null) Items.BLACK_BACKGROUND else deck[i]!!.clone()
            val turn = turns[i]

            if (player.game.difficulty.suggestions && turn?.goodChoice(player) ?: false)
                item.editMeta { it.setEnchantmentGlintOverride(true) }

            inv.setItem(i, item)
        }

        inv.setItem(13, ItemBuilder((if (coins == 0) Material.RESIN_BRICK else Material.GOLD_INGOT)).apply {
            amount(coins.coerceAtLeast(1))
            name(player.locale().component("item-lore.cost.value", coins.toString()))
        }.build())
    }
}
